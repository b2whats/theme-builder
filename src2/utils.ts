import type { ThemeBuilder } from './ThemeBuilder'

export const isObject = (value: any): value is object => value && value.constructor && value.constructor === Object
export const isFunction = (value: any): value is Function => typeof value === 'function'
export const isIterable = (value: any): value is object => Object(value) === value
export const isThemeBuilder = (value: any): value is ThemeBuilder<any> => value && value.constructor && value.constructor.name === 'ThemeBuilder'

export function get(
  obj: object,
  path: string | number,
) {
  const key = typeof path === "string" ? path.split(".") : [path]

  for (let index = 0; index < key.length; index += 1) {
    if (!obj) break
    obj = obj[key[index]]
  }

  return obj
}

export const flatObjectEntries = <
  O extends Record<string, any>
>(object: O, delimiter: string = '.', prefix: string = ''): [string, any][] => Object
  .entries(object)
  .flatMap(
    ([key, value]) => isIterable(value)
      ? flatObjectEntries(value, delimiter, `${prefix}${key}${delimiter}`)
      : [[ `${prefix}${key}`, value ]]
  )

type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...0[]]

type FlatObjectEntries<O, Delimiter extends string = '.', Prefix extends string = '', Key extends keyof O = keyof O, I extends number= 5> = [I] extends [never] ? never : Key extends string | number
  ? O[Key] extends Record<string, any>
    ? FlatObjectEntries<O[Key], Delimiter, `${Prefix}${Key}${Delimiter}`, Exclude<keyof O[Key], keyof any[]>, Prev[I]>
    : [`${Prefix}${Key}`, O[Key]]
  : never

type TupleEntrie<T, I extends number> = T extends [...any[]] ? T[I] : never

export type FlatObjectReturn<O extends {}, Delimiter extends string> = FlatObjectEntries<O, Delimiter> extends infer Entries
  ? { [K in Entries as TupleEntrie<K, 0>]: TupleEntrie<K, 1> }
  : never

export const flatObject = <
  O extends Record<string, any>,
  D extends string = '.'
>(object: O, delimiter?: D): FlatObjectReturn<O, D> => Object.fromEntries(flatObjectEntries(object, delimiter || '.')) as any

type Cast<A, B> = A extends B ? A : B

type Narrowable =
| string
| number
| bigint
| boolean

export type Narrow<A> = Cast<A,
  | (A extends [] ? [] : never)
  | (A extends Narrowable ? A : never)
  | ({ [K in keyof A]: A[K] extends Function ? A[K] : Narrow<A[K]> })
>

type FlattenType<T extends Record<string, any>> = T extends unknown ? T : never

type LeavesEntries<
  O,
  Delimiter extends string = '.',
  Prefix extends string = '',
  Key extends keyof O = keyof O,
  I extends number = 5
> = [I] extends [never] ? never : Key extends string | number
  ? O[Key] extends readonly any[] | any[]
    ? LeavesEntries<O[Key], Delimiter, `${Prefix}${Key}${Delimiter}`, Exclude<keyof O[Key], keyof any[]>, Prev[I]>
  : O[Key] extends Record<string, any>
    ? LeavesEntries<O[Key], Delimiter, `${Prefix}${Key}${Delimiter}`, Exclude<keyof O[Key], Exclude<keyof any[], number>>, Prev[I]>
    : [`${Prefix}${Key}`, CastToPrimitive<O[Key]>]
  : never


type CastToPrimitive<T> =
  T extends string ? string :
  T extends number ? number :
  T extends any ? T :
  never

export type PathsEntries<
  O,
  Delimiter extends string = '.',
  Prefix extends string = '',
  Key extends keyof O = keyof O,
  I extends number = 5
> = [I] extends [never] ? never : Key extends string | number ?
  O[Key] extends readonly any[] | any[] ?
    | PathsEntries<O[Key], Delimiter, `${Prefix}${Key}${Delimiter}`, Exclude<keyof O[Key], keyof any[]>, Prev[I]>
    | [`${Prefix}${Key}`, Exclude<keyof O[Key], keyof any[]>] :
  O[Key] extends Record<string, any> ?
    | [`${Prefix}${Key}`, keyof O[Key]]
    | PathsEntries<O[Key], Delimiter, `${Prefix}${Key}${Delimiter}`, keyof O[Key], Prev[I]> :
  [`${Prefix}${Key}`, CastToPrimitive<O[Key]>] : never

export type ObjectPaths<O extends Record<string, any>, Entries extends [string, any] = PathsEntries<O>> = FlattenType<{ [K in Entries as K[0] ]: K[1] }>
export type ObjectLeaves<O extends Record<string, any>, Entries extends [string, any] = LeavesEntries<O>> = { [K in Entries as K[0] ]: K[1] }

type UnionToIntersection<U> =   (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never
type UnionToFunctions<U> = U extends unknown ? (k: U) => void : never

type IntersectionOf<T> =
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; (d: infer D): void; (e: infer E): void; (f: infer F): void; (g: infer G): void; (h: infer H): void; } ? [A, B, C, D, E, F, G, H] :
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; (d: infer D): void; (e: infer E): void; (f: infer F): void; (g: infer G): void; } ? [A, B, C, D, E, F, G] :
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; (d: infer D): void; (e: infer E): void; (f: infer F): void; } ? [A, B, C, D, E, F] :
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; (d: infer D): void; (e: infer E): void; } ? [A, B, C, D, E] :
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; (d: infer D): void; } ? [A, B, C, D] :
  T extends { (a: infer A): void; (b: infer B): void; (c: infer C): void; } ? [A, B, C] :
  T extends { (a: infer A): void; (b: infer B): void; } ? [A, B] :
  T extends { (a: infer A): void } ? [A] :
  never

export type UnionToTuple<T> = IntersectionOf<UnionToIntersection<UnionToFunctions<T>>>