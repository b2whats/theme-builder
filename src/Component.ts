import { css } from '@emotion/css'
import { mergeTheme } from './merge'
import type { Properties } from './Properties'
import type { DeepPartial, MaybeArray, NoInfer, ComponentType } from './utils'
import { objectHash, isEmptyObject } from './utils'

type OrFunction<Args, Return> = (props: Args) => Return
type AddPropertyFunction<L, Props> = {
  [K in keyof L]?: L[K] extends Record<string, any> ? AddPropertyFunction<L[K], Props> : L[K] | OrFunction<Props, L[K]>
}

type AdditionalProps<Props> = {
  withProps?: boolean | ((props: Props) => Partial<Props>)
  className?: string
  asString?: string
}

type SimpleComponent<P = any> = ComponentType<P, JSX.Element>
type CurryComponent<C = any, P = any> = (cond: C) => SimpleComponent<P>
type ChildrenBlock<ChildrenProps, StyleProps, Props> = {
  children?: {
    component?: CurryComponent<Props, ChildrenProps> | SimpleComponent<ChildrenProps>
  } & (ChildrenProps extends object ? AddPropertyFunction<ChildrenProps, Props> : Partial<StyleProps>)
}

type Flatten<T> = T extends unknown ? T : never

type PropertyConfig<List, Props, ChildrenProps> = List extends Properties<any, infer L, any> ? AddPropertyFunction<L, Props> & AdditionalProps<Props> & ChildrenBlock<ChildrenProps, L, Props> : never
type Slot<Name extends string, List> = {
  [key in Name]: List
}

export type StyleSlots<Slots> = {
  [K in keyof Slots]: {
    className: string,
    children: Slots[K] extends { children: { component: infer C, [key: string]: any } }
      ? C extends CurryComponent ? ReturnType<C> : C
      : undefined
    childrenProps:  Slots[K] extends { children: infer C } ? Flatten<Pick<C, Exclude<keyof C, 'component'>>> : undefined,
  }
}

type NonArrayObject = {
  [x: string]: any
  [y: number]: never
}

type Validate<T, U> = U & {
  [K in keyof T]: K extends Extract<keyof T, keyof U>
    ? (T[K] extends NonArrayObject ? Validate<T[K], NonNullable<U[K]>> : T[K])
    : `Not possible property key - ${K & string}`
}

type ComponentTheme<Props> = {
  defaultProps?: Partial<Props>
  mapProps?: (props: Props) => Partial<Props>
  slots: {
    [key: string]: AdditionalProps<Props> & ChildrenBlock<any, {}, any>
  }
}
type ComponentThemeAsList<Name extends string, Props> = {
  [key in Name]?: Partial<ComponentTheme<Props>>
}
export class Component<
  Name extends string = never,
  Props = never,
  Slots extends Record<string, any> = {},
  PropertiesList extends Properties<any, any> = never
> {
  componentName!: string
  theme: ComponentTheme<Props> = {
    defaultProps: undefined,
    mapProps: undefined,
    slots: {},
  }
  private styleCache: Map<object, Map<string, object>> = new Map()
  private cache: Map<object, object> = new Map()

  constructor(public properties: PropertiesList) {}

  types<P extends Record<string, any>>(): Component<Name, P, Slots, PropertiesList> {
    return this as any
  }

  name<N extends string>(name: N): Component<N, Props, Slots, PropertiesList> {
    this.componentName = name

    return this 
  }

  defaultProps(props: Partial<Props>) {
    this.theme.defaultProps = props

    return this
  }

  mapProps(map: (props: Props) => Partial<Props>) {
    this.theme.mapProps = map

    return this
  }

  slot<
    SlotName extends string,
    Config,
    ChildrenProps,
  >(
    name: SlotName,
    config: Validate<Config, PropertyConfig<PropertiesList, Props, ChildrenProps>>
  ): Component<Name, Props, Slots & Slot<SlotName, Config>, PropertiesList> {
    this.theme.slots[name] = config
    
    return this
  }

  objectStyleToString(tokens: DeepPartial<this['properties']['tokens']>, props: Props, rules: {}): string {
    let css = ''
    const responsiveCss: string[] = []

    if (rules['asString']) {
      css += rules['asString']
    }

    for (const rule in rules) {
      let value = rules[rule]
      const type = typeof value

      if (!this.properties.rules[rule]) continue

      if (type === 'function') {
        value = value(props)
      }

      if (Array.isArray(value)) {
        for (let index = 0; index < value.length; index++) {
          responsiveCss[index] = this.properties.compute(rule, value[index], tokens)
        }
        continue
      }

      if (type === 'object') {
        value = this.objectStyleToString(tokens, props, value)
      }

      css += this.properties.compute(rule, value, tokens)
    }

    if (responsiveCss.length) {
      const responsiveRules = this.properties.breakpointsRule(tokens)
      css += responsiveCss[0]

      for (let index = 1; index < responsiveCss.length; index++) {
        css += responsiveCss[index] !== '' ? `${responsiveRules[index]} {${responsiveCss[index]}}` : ''
      }
    }

    return css
  }

  mergeTheme(theme: Record<string, any>) {
    let themeCache = this.cache.get(theme)

    if (themeCache === undefined) {
      const componentTheme = {
        ...this.properties.tokens,
        [this.componentName]: this.theme
      }
      themeCache = mergeTheme(componentTheme, theme)
      this.cache.set(theme, themeCache)
    }

    return themeCache
  }

  execute(theme: ComponentThemeAsList<Name, Props> & DeepPartial<this['properties']['tokens']>, props: Props): StyleSlots<Slots> {
    let styleCache = this.styleCache.get(theme)

    if (styleCache === undefined) {
      styleCache = new Map()
      this.styleCache.set(theme, styleCache)
    }

    const propsHash = objectHash(props)
    let propsCache = styleCache.get(propsHash)

    if (propsCache) return propsCache as StyleSlots<Slots>

    console.log('execute')

    const cacheTheme = this.mergeTheme(theme) as DeepPartial<this['properties']['tokens']>
    const componentTheme = cacheTheme[this.componentName] as this['theme'] 

    props = {
      ...componentTheme.defaultProps,
      ...props,
    }

    if (componentTheme.mapProps) Object.assign(props, componentTheme.mapProps(props))

    const result  = {} as StyleSlots<Slots>
    
    for (let [name, { withProps, className, children: { component, ...childrenProps } = { component: undefined }, ...rules }] of Object.entries(componentTheme.slots)) {
      if (withProps) {
        Object.assign(rules, typeof withProps === 'function' ? withProps(props) : props)
      }

      result[name as keyof Slots] = {
        className: (className ? className + ' ' : '') + css(this.objectStyleToString(cacheTheme, props, rules)),
        children: component,
        childrenProps: isEmptyObject(childrenProps) ? undefined : childrenProps,
      } as any
    }

    styleCache.set(propsHash, result)
    return result
  }
}

type SwitchMap<Props, Prop extends keyof Props, Result> = {
  [Value in Extract<Props[Prop], string>]?: Result | ((props: Props) => Result | undefined)
}
interface SlotUtils {
  if<ExternalProps extends Record<string, any>, Props extends keyof ExternalProps, Values>(
    prop: MaybeArray<Props>,
    value: NoInfer<Values>,
    defaultValue?: NoInfer<Values>,
  ): (props: ExternalProps) => Values
  map<ExternalProps, Prop extends keyof ExternalProps, Values>(
    prop: Prop,
    body: SwitchMap<ExternalProps, Prop, Values>,
    defaultValue?: NoInfer<Values>,
  ): (props: ExternalProps) => Values
}

export const utils: SlotUtils = {
  if: (keys, value, defaultValue) => {
    return function utils(props: any): any {
      const isValue = Array.isArray(keys)
        ? keys.every(key => !!props[key])
        : props[keys]
      
      return isValue ? value : defaultValue
    }
  },
  map: (prop, options, defaultValue) => {
    return function utils(props: any) {
      const value = props[prop]

      return (typeof options[value] === 'function' ? options[value](props) : options[value]) || defaultValue
    }
  },
}


