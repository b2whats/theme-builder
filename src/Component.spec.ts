import { properties, tokens, component } from './fixtures'
import { Component } from './Component'
import { literalValues } from './utils'

expect.addSnapshotSerializer({
  test: (val) => typeof val === 'function',
  print: (val) => `${val}`
})


const theme = {}
test('Component merge theme - none', () => {
  const empty = {}
  const notSame = {}
  const theme = component.mergeTheme(empty)
  const theme2 = component.mergeTheme(empty)
  const theme3 = component.mergeTheme(notSame)

  expect(theme).toMatchSnapshot()
  expect(theme === theme2).toBe(true)
  expect(theme !== theme3).toBe(true)
})



test('Component merge theme - all', () => {
  const componentTheme = literalValues({
    Button: {
      defaultProps: { size: 'm'},
      mapProps: (props: any) => ({ size: 'l' as 'l' }),
      slots: {
        base: {
          as: 'div',
          display: 'flex',
          color: ['primary', 'red'],
          fontSize: (props: any) => props.otherSize,
          hover: {
            display: 'flex',
            color: ['primary', 'red'],
            fontSize: (props: any) => props.otherSize,
          }
        }
      }
    }
  })
  const theme = component.mergeTheme(componentTheme)
  const theme2 = component.mergeTheme(componentTheme)
  const theme3 = component.mergeTheme({...componentTheme})

  expect(theme === theme2).toBe(true)
  expect(theme !== theme3).toBe(true)

  expect(theme).toMatchSnapshot()
})
test('Component merge theme - undefined', () => {
  const componentTheme = literalValues({
    Button: {
      defaultProps: undefined,
      mapProps: undefined,
      slots: {
        base: {
          as: 'div',
          display: undefined,
          color: undefined,
          fontSize: undefined,
          hover: {
            display: undefined,
            color: undefined,
            fontSize: undefined,
          }
        }
      }
    }
  })
  const theme = component.mergeTheme(componentTheme)

  expect(theme).toMatchSnapshot()
})


test('Component execute', () => {
  const props = literalValues({
    size: 's'
  })
  const theme = {}
  const result = component.execute(theme, props)

  expect(result).toMatchSnapshot()
})

test('Component execute array', () => {
  type Props = {
    children?: any
    size?: 's' | 'm' | 'l'
    name?: string
    value?: string | number
    kind?: 'fill' | 'outline' | 'flat'
    checked?: boolean
    type?: 'button' | 'submit' | 'reset'
    loading?: boolean
    shape?: 'pill' | 'square' | 'circle'
    href?: string
  }
  
  const props: Props = {
    size: 'l',
    name: 'sbmt',
    kind: 'fill',
    type: 'button',
    shape: 'pill',
  }

  const arrayComponent = new Component(properties)
  .name('Button')
  .types<Props>()
  .defaultProps({
    size: 's',
    kind: 'flat',
    checked: false,
  })
  .mapProps((props) => ({
    size: 'm'
  }))
  .slot('button', {
    display: ['block', 'flex'],
    color: ['black', 'black12'],
    fontSize: ['s', 's'],
    lineHeight: ['none', 'dense'],
    italic: [true, false],
    noWrap: [true, false],
    uppercase: [true, false],
    cursor: ['pointer', 'disabled'],
    smoothing: ['antialiased', 'auto'],
    borderRadius: ['l', 'm'],
    column: [true, false],
    hover: {
      display: ['block', 'flex'],
      color: ['black', 'black12'],
      fontSize: ['s', 's'],
      lineHeight: ['none', 'dense'],
      italic: [true, false],
      noWrap: [true, false],
      uppercase: [true, false],
      cursor: ['pointer', 'disabled'],
      smoothing: ['antialiased', 'auto'],
      borderRadius: ['l', 'm'],
      column: [true, false],
    },
    active: {
      display: ['block', 'flex'],
      color: ['black', 'black12'],
      fontSize: ['s', 's'],
      lineHeight: ['none', 'dense'],
      italic: [true, false],
      noWrap: [true, false],
      uppercase: [true, false],
      cursor: ['pointer', 'disabled'],
      smoothing: ['antialiased', 'auto'],
      borderRadius: ['l', 'm'],
      column: [true, false],
    }
  })

  expect(arrayComponent.execute({}, props)).toMatchSnapshot()
})