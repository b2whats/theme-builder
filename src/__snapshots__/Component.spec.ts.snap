// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component execute 1`] = `
"===========base===========
.css-1cwwvg5 {
  color: red;
  display: block;
  font-size: 36px;
  color: rgba(0, 0, 0, 1);
}

.css-1cwwvg5.css-1cwwvg5:hover {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media screen and (min-width: 52em) {
  .css-1cwwvg5.css-1cwwvg5:hover {
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
  }
}

.css-1cwwvg5::before {
  color: rgba(0, 0, 0, 1);
}

@media screen and (min-width: 52em) {
  .css-1cwwvg5::before {
    color: #E6F6FF;
  }
}

@media screen and (min-width: 52em) {
  .css-1cwwvg5 {
    color: #E6F6FF;
  }
}

"
`;

exports[`Component execute 2`] = `
"===========base2===========
.css-1dxbwyp {
  color: red;
  display: inline;
  color: rgba(0, 0, 0, 0.04);
}

.css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover:disabled {
  font-size: 16px;
  display: inline;
  color: rgba(0, 0, 0, 0.04);
}

@media screen and (min-width: 52em) {
  .css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover.css-1dxbwyp.css-1dxbwyp:hover:disabled {
    color: #FFEDEE;
  }
}

@media screen and (min-width: 52em) {
  .css-1dxbwyp {
    color: rgba(0, 0, 0, 0.04);
  }
}

"
`;

exports[`Component execute array 1`] = `
Object {
  "button": Object {
    "children": undefined,
    "childrenProps": undefined,
    "className": "css-ody2sg",
  },
}
`;

exports[`Component merge theme - all 1`] = `
Object {
  "Button": Object {
    "defaultProps": Object {
      "size": "m",
    },
    "mapProps": function (props) { return (__assign(__assign({}, a(props)), b(props))); },
    "slots": Object {
      "base": Object {
        "&::before": Object {
          "color": Array [
            "black",
            "blue50",
          ],
        },
        "as": "div",
        "asString": "color: red;",
        "color": Array [
          "primary",
          "red",
        ],
        "display": "flex",
        "fontSize": function (props) { return props.otherSize; },
        "hover": Object {
          "color": Array [
            "primary",
            "red",
          ],
          "display": "flex",
          "fontSize": function (props) { return props.otherSize; },
          "grow": Array [
            true,
            false,
          ],
        },
      },
      "base2": Object {
        "asString": "color: red;",
        "children": Object {
          "component": function (props) { return props.size === 'm' ? C1 : C; },
          "kind2": "fill",
        },
        "color": Array [
          "black4",
          "black4",
        ],
        "display": function (props) { return 'inline'; },
        "hover": Object {
          "disabled": Object {
            "color": Array [
              "black4",
              "red50",
            ],
            "display": Array [
              "inline",
            ],
            "fontSize": function (props) { return 'm'; },
          },
        },
      },
    },
  },
  "breakpoints": Array [
    "40em",
    "52em",
    "64em",
    "80em",
  ],
  "dimension": Object {
    "borderRadius": Object {
      "l": 5,
      "m": 3,
      "s": 1,
    },
    "rowHeight": Object {
      "l": 48,
      "m": 36,
      "s": 28,
    },
  },
  "focus": "white 0px 0px 0px 1px, 0px 0px 2px 3px black",
  "font": Object {
    "fontSize": Object {
      "l": 22,
      "m": 16,
      "s": 13,
      "xl": 26,
      "xs": 11,
      "xxl": 36,
      "xxxl": 0,
      "xxxxl": 0,
      "xxxxxl": 0,
    },
    "lineHeight": Object {
      "dense": 1.25,
      "none": 1.12,
      "normal": 1.375,
    },
  },
  "opacity": Object {
    "minLevel": 0,
    "op1": 1,
  },
  "palette": Object {
    "black": "rgba(0, 0, 0, 1)",
    "black12": "rgba(0, 0, 0, 0.12)",
    "black4": "rgba(0, 0, 0, 0.04)",
    "black8": "rgba(0, 0, 0, 0.08)",
    "blue100": "#CCECFF",
    "blue200": "#A1DFFF",
    "blue300": "#80D4FF",
    "blue50": "#E6F6FF",
    "gray12": "#E0E0E0",
    "gray16": "#D6D6D6",
    "gray4": "#F5F5F5",
    "gray8": "#EBEBEB",
    "gray96": "#0a0a0a",
    "green100": "#EEF7DC",
    "green200": "#CBE792",
    "green300": "#C1E27E",
    "green50": "#EEF7DC",
    "orange100": "#FFE7BA",
    "orange200": "#FFD78F",
    "orange300": "#FFE27A",
    "orange50": "#FFF5E2",
    "purple100": "#F0E7FE",
    "purple200": "#DAC5FB",
    "purple300": "#CAACFA",
    "purple50": "#F3EDFE",
    "red100": "#FFE6E6",
    "red200": "#FFC5C6",
    "red300": "#FFACAD",
    "red50": "#FFEDEE",
    "white": "#FFFFFF",
    "yellow100": "#fff3d9",
    "yellow200": "#ffe5b2",
    "yellow300": "#ffd173",
    "yellow50": "#fff9ec",
  },
  "shadow": Object {
    "1": "1",
    "2": "2",
    "3": "3",
    "4": "4",
  },
  "space": Object {
    "l": 20,
    "m": 16,
    "s": 12,
    "xl": 24,
    "xs": 8,
    "xxl": 32,
    "xxs": 4,
    "xxxl": 40,
    "xxxs": 2,
  },
  "text": Object {
    "h1": Object {
      "fontSize": 12,
      "lineHeight": 1.5,
    },
    "h2": Object {
      "fontSize": 15,
      "lineHeight": 1.7,
    },
  },
  "transition": Object {
    "fade": "",
  },
  "zIndex": Object {
    "dropdown": 1,
    "popup": 1,
    "tooltip": 1,
  },
}
`;

exports[`Component merge theme - none 1`] = `
Object {
  "Button": Object {
    "defaultProps": Object {
      "size": "s",
    },
    "mapProps": function (props) { return ({
    size: 'm',
}); },
    "slots": Object {
      "base": Object {
        "&::before": Object {
          "color": Array [
            "black",
            "blue50",
          ],
        },
        "asString": "color: red;",
        "color": Array [
          "black",
          "blue50",
        ],
        "display": function (props) { return 'block'; },
        "fontSize": "xxl",
        "hover": Object {
          "grow": Array [
            true,
            false,
          ],
        },
      },
      "base2": Object {
        "asString": "color: red;",
        "children": Object {
          "component": function (props) { return props.size === 'm' ? C1 : C; },
          "kind2": "fill",
        },
        "color": Array [
          "black4",
          "black4",
        ],
        "display": function (props) { return 'inline'; },
        "hover": Object {
          "disabled": Object {
            "color": Array [
              "black4",
              "red50",
            ],
            "display": Array [
              "inline",
            ],
            "fontSize": function (props) { return 'm'; },
          },
        },
      },
    },
  },
  "breakpoints": Array [
    "40em",
    "52em",
    "64em",
    "80em",
  ],
  "dimension": Object {
    "borderRadius": Object {
      "l": 5,
      "m": 3,
      "s": 1,
    },
    "rowHeight": Object {
      "l": 48,
      "m": 36,
      "s": 28,
    },
  },
  "focus": "white 0px 0px 0px 1px, 0px 0px 2px 3px black",
  "font": Object {
    "fontSize": Object {
      "l": 22,
      "m": 16,
      "s": 13,
      "xl": 26,
      "xs": 11,
      "xxl": 36,
      "xxxl": 0,
      "xxxxl": 0,
      "xxxxxl": 0,
    },
    "lineHeight": Object {
      "dense": 1.25,
      "none": 1.12,
      "normal": 1.375,
    },
  },
  "opacity": Object {
    "minLevel": 0,
    "op1": 1,
  },
  "palette": Object {
    "black": "rgba(0, 0, 0, 1)",
    "black12": "rgba(0, 0, 0, 0.12)",
    "black4": "rgba(0, 0, 0, 0.04)",
    "black8": "rgba(0, 0, 0, 0.08)",
    "blue100": "#CCECFF",
    "blue200": "#A1DFFF",
    "blue300": "#80D4FF",
    "blue50": "#E6F6FF",
    "gray12": "#E0E0E0",
    "gray16": "#D6D6D6",
    "gray4": "#F5F5F5",
    "gray8": "#EBEBEB",
    "gray96": "#0a0a0a",
    "green100": "#EEF7DC",
    "green200": "#CBE792",
    "green300": "#C1E27E",
    "green50": "#EEF7DC",
    "orange100": "#FFE7BA",
    "orange200": "#FFD78F",
    "orange300": "#FFE27A",
    "orange50": "#FFF5E2",
    "purple100": "#F0E7FE",
    "purple200": "#DAC5FB",
    "purple300": "#CAACFA",
    "purple50": "#F3EDFE",
    "red100": "#FFE6E6",
    "red200": "#FFC5C6",
    "red300": "#FFACAD",
    "red50": "#FFEDEE",
    "white": "#FFFFFF",
    "yellow100": "#fff3d9",
    "yellow200": "#ffe5b2",
    "yellow300": "#ffd173",
    "yellow50": "#fff9ec",
  },
  "shadow": Object {
    "1": "1",
    "2": "2",
    "3": "3",
    "4": "4",
  },
  "space": Object {
    "l": 20,
    "m": 16,
    "s": 12,
    "xl": 24,
    "xs": 8,
    "xxl": 32,
    "xxs": 4,
    "xxxl": 40,
    "xxxs": 2,
  },
  "text": Object {
    "h1": Object {
      "fontSize": 12,
      "lineHeight": 1.5,
    },
    "h2": Object {
      "fontSize": 15,
      "lineHeight": 1.7,
    },
  },
  "transition": Object {
    "fade": "",
  },
  "zIndex": Object {
    "dropdown": 1,
    "popup": 1,
    "tooltip": 1,
  },
}
`;

exports[`Component merge theme - undefined 1`] = `
Object {
  "Button": Object {
    "slots": Object {
      "base": Object {
        "&::before": Object {
          "color": Array [
            "black",
            "blue50",
          ],
        },
        "as": "div",
        "asString": "color: red;",
        "hover": Object {
          "grow": Array [
            true,
            false,
          ],
        },
      },
      "base2": Object {
        "asString": "color: red;",
        "children": Object {
          "component": function (props) { return props.size === 'm' ? C1 : C; },
          "kind2": "fill",
        },
        "color": Array [
          "black4",
          "black4",
        ],
        "display": function (props) { return 'inline'; },
        "hover": Object {
          "disabled": Object {
            "color": Array [
              "black4",
              "red50",
            ],
            "display": Array [
              "inline",
            ],
            "fontSize": function (props) { return 'm'; },
          },
        },
      },
    },
  },
  "breakpoints": Array [
    "40em",
    "52em",
    "64em",
    "80em",
  ],
  "dimension": Object {
    "borderRadius": Object {
      "l": 5,
      "m": 3,
      "s": 1,
    },
    "rowHeight": Object {
      "l": 48,
      "m": 36,
      "s": 28,
    },
  },
  "focus": "white 0px 0px 0px 1px, 0px 0px 2px 3px black",
  "font": Object {
    "fontSize": Object {
      "l": 22,
      "m": 16,
      "s": 13,
      "xl": 26,
      "xs": 11,
      "xxl": 36,
      "xxxl": 0,
      "xxxxl": 0,
      "xxxxxl": 0,
    },
    "lineHeight": Object {
      "dense": 1.25,
      "none": 1.12,
      "normal": 1.375,
    },
  },
  "opacity": Object {
    "minLevel": 0,
    "op1": 1,
  },
  "palette": Object {
    "black": "rgba(0, 0, 0, 1)",
    "black12": "rgba(0, 0, 0, 0.12)",
    "black4": "rgba(0, 0, 0, 0.04)",
    "black8": "rgba(0, 0, 0, 0.08)",
    "blue100": "#CCECFF",
    "blue200": "#A1DFFF",
    "blue300": "#80D4FF",
    "blue50": "#E6F6FF",
    "gray12": "#E0E0E0",
    "gray16": "#D6D6D6",
    "gray4": "#F5F5F5",
    "gray8": "#EBEBEB",
    "gray96": "#0a0a0a",
    "green100": "#EEF7DC",
    "green200": "#CBE792",
    "green300": "#C1E27E",
    "green50": "#EEF7DC",
    "orange100": "#FFE7BA",
    "orange200": "#FFD78F",
    "orange300": "#FFE27A",
    "orange50": "#FFF5E2",
    "purple100": "#F0E7FE",
    "purple200": "#DAC5FB",
    "purple300": "#CAACFA",
    "purple50": "#F3EDFE",
    "red100": "#FFE6E6",
    "red200": "#FFC5C6",
    "red300": "#FFACAD",
    "red50": "#FFEDEE",
    "white": "#FFFFFF",
    "yellow100": "#fff3d9",
    "yellow200": "#ffe5b2",
    "yellow300": "#ffd173",
    "yellow50": "#fff9ec",
  },
  "shadow": Object {
    "1": "1",
    "2": "2",
    "3": "3",
    "4": "4",
  },
  "space": Object {
    "l": 20,
    "m": 16,
    "s": 12,
    "xl": 24,
    "xs": 8,
    "xxl": 32,
    "xxs": 4,
    "xxxl": 40,
    "xxxs": 2,
  },
  "text": Object {
    "h1": Object {
      "fontSize": 12,
      "lineHeight": 1.5,
    },
    "h2": Object {
      "fontSize": 15,
      "lineHeight": 1.7,
    },
  },
  "transition": Object {
    "fade": "",
  },
  "zIndex": Object {
    "dropdown": 1,
    "popup": 1,
    "tooltip": 1,
  },
}
`;
